Use sorted set in Authorizations

DomGarguilo opened the pull request:
Fixes #58

keith-turner commented on src/main/java/org/apache/accumulo/access/Authorizations.java:
```diff
@@ -33,13 +36,13 @@
   }
 
-  public Set<String> asSet() {
+  public SortedSet<String> asSet() {
```
This returns a mutable object that will change the internal state of the object when its mutated.  The javadoc states its immutable.

keith-turner reviewed (commented):
I made some comments on #58 about the general direction of this.  IMO we should do the following.
  
 * Sort in toString for nicer output
 * Keep Set instead of SortedSet in the impl and API.  Also keep using Set.copyOf.
 * Add a unit test to ensure that what asSet() returns is immutable.  Try to add something and verify an exception is thrown and the source did not change.

DomGarguilo commented:
> I made some comments on #58 about the general direction of this. IMO we should do the following.
> 
> * Sort in toString for nicer output
> * Keep Set instead of SortedSet in the impl and API.  Also keep using Set.copyOf.
> * Add a unit test to ensure that what asSet() returns is immutable.  Try to add something and verify an exception is thrown and the source did not change.

Addressed these points in 6aee3e1.

keith-turner commented on src/main/java/org/apache/accumulo/access/Authorizations.java:
```diff
@@ -58,9 +59,12 @@ public int hashCode() {
   @Override
   public String toString() {
-    return authorizations.toString();
+    return authorizations.stream().sorted().collect(Collectors.joining(","));
```
The set to toString will put `[` and `]` around data which is nice.   Could possibly add those.

dlmarion replied:
I have no issue with this, but I'm curious on the difference between this and `new TreeSet<>(authorizations).toString()`

DomGarguilo replied:
Using a TreeSet will put a space after each comma and add the brackets. e.g. `[A, B, Y, Z]`

Using the stream with `.joining(",")` will not add the spaces nor the brackets by default. e.g. `A,B,Y,Z` but the spaces and brackets can be added obviously.

The TreeSet code would be more simple I think but also would have the overhead of creating the new TreeSet (which might be negligible in this case).

keith-turner replied:
Creating a treeset seems fine for this case.  Thinking is that when toString is called its for the purpose of showing a person something, so in that case the expense is worthwhile.  Can optimize later if ever needed.

dlmarion replied:
> The TreeSet code would be more simple I think but also would have the overhead of creating the new TreeSet

I think that assumes that the streaming example creates no objects. 

`Collection.stream` calls `StreamSupport.stream`, which creates a new `ReferencePipeline.Head`:
https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/util/stream/StreamSupport.java#L67

Calling `sorted` on the `ReferencePipeline` creates more objects:
https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/util/stream/ReferencePipeline.java#L462

`Collectors.joining` creates two new objects.

https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/util/stream/Collectors.java#L367

DomGarguilo replied:
Made toString use TreeSet in 51e02f5

dlmarion commented on src/main/java/org/apache/accumulo/access/Authorizations.java:
```diff
@@ -58,9 +59,12 @@ public int hashCode() {
   @Override
   public String toString() {
-    return authorizations.toString();
+    return new TreeSet<>(authorizations).toString();
```
I wonder what the penalty is here when the user doesn't care about sorting? Consider the example:
```
  LOG.info("Authorizations: {}", authorizations);
```

I wonder if we should leave `toString()` the way that it was, and add a new method with this implementation called `toSortedString` or something. i'm usually on the fence regarding changes like this because we are baking in a performance penalty 100% of the time for something that we *think* is a good idea. I'd rather let the user choose if they want that penalty.

DomGarguilo replied:
I'm fine with that. Although in that case do you think its worth even creating a new method like `toSortedString` or should we just let the user use `toSet` and sort and create the string on their own?

dlmarion replied:
I'm fine with letting users do it on their own. We should get the issue originator to weigh in on this.

DomGarguilo replied:
@ctubbsii do you have any input on this as the creator of the issue associated with this PR?

ctubbsii replied:
I think now that Authorizations.of() takes a set, we could just use the set that the user passed in, preserving whatever order it already had. If it's an immutable set already, we can just keep a reference to the same object instead of doing an unnecessary copy. If it's not, we can preserve the order when we do the copy (maybe using `LinkedHashSet.addAll()` wrapped with `Collections.unmodifiableSet`?). The only issue with this approach is that `asSet()` won't return something that has the `SortedSet` interface if it originally was a mutable sorted set, like TreeSet, but that might be an acceptable compromise).
